File: classes03/cpp/webcamStabilizedFLD.cpp
/**
 * @brief 使用光流法增加追踪的稳定性
 * 
 * @file webcamStabilizedFLDcpp
 * @author 张晓民
 * @date 2018-08-20
 */
#include <dlib/opencv.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing/render_face_detections.h>
#include <dlib/image_processing.h>
#include <dlib/gui_widgets.h>
#include <opencv2/opencv.hpp>
#include "renderFace.hpp"

using namespace dlib;
using namespace std;
using namespace cv;

#define RESIZE_HEIGHT 360
#define SKIP_FRAMES 1
#define OPENCV_FACE_RENDER
#define NUM_FRAMES_FOR_FPS 100

void constrainPoint(cv::Point2f &p, cv::Size sz)
{
    p.x = cv::min(cv::max( (double)p.x,0.0), (double)(sz.width - 1));
    p.y = cv::min(cv::max( (double)p.y,0.0), (double)(sz.height - 1));
}

double interEyeDistance(dlib::full_object_detection &shape)
{
    cv::Point2f leftEyeLeftCorner(shape.part(36).x(), shape.part(36).y());
	cv::Point2f rightEyeRightCorner(shape.part(45).x(), shape.part(45).y());
	double distance = norm(rightEyeRightCorner - leftEyeLeftCorner); 
	return distance;
}

int main(void)
{
    // 创建一个命名window来显示结果
    string winName("Stabilized Facial Landmark Detector");
    cv::namedWindow(winName, cv::WINDOW_NORMAL);

    // 打开摄像头
    cv::VideoCapture cap(0);
    if(!cap.isOpened()) // 检查摄像头是否成功打开
    {
        std::cout<<"摄像头打开失败!"<<std::endl;
        return -1;
    }
    
    std::cout<<"摄像头打开成功！"<<std::endl;
    
    // 设置光流法的参数
    cv::TermCriteria termcrit(cv::TermCriteria::COUNT|cv::TermCriteria::EPS,20,0.03);
    cv::Size winSize(101,101);
    double eyeDistance, dotRadius, sigma;
    bool eyeDistanceNotCalculated = true;
    int maxLevel = 5;
    std::vector<uchar> status;
    std::vector<float> err;


    double fps=30.0;

    // 当前帧、上一帧及其灰度版本
    cv::Mat im, imPrev, imGray, imGrayPrev;
    // 图像金字塔
    std::vector<cv::Mat> imGrayPyr, imGrayPrevPyr;

    // 获取第一帧
    cap >> imPrev;

    // 转换成灰度图像
    cv::cvtColor(imPrev, imGrayPrev, cv::COLOR_BGR2GRAY);

    // 构建图像金字塔
    cv::buildOpticalFlowPyramid(imGrayPrev, imGrayPrevPyr, winSize, maxLevel);

    // 获取图像的尺寸
    cv::Size size=imPrev.size();

    // 使用固定高度的图片作为face detector的输入图像
    cv::Mat imSmall, imDisplay;
    float height=imPrev.rows;
    std::cout<<"摄像头原始图像的高度为:"<<height<<std::endl;

    // 对原始图像进行处理
    float RESIZE_SCALE= height/RESIZE_HEIGHT;
    // float RESIZE_SCALE= 1.0;
    

    dlib::frontal_face_detector detector=get_frontal_face_detector();
    dlib::shape_predictor predictor;
    dlib::deserialize("../../common/resources/shape_predictor_68_face_landmarks.dat") >> predictor;

    // 初始化计时器
    double t = (double)cv::getTickCount();
    int count=0;

    std::vector<dlib::rectangle> faces;

    // 创建一些变量来存储光流法需要的数据
    std::vector<cv::Point2f> points, pointsPrev, pointsDetectedCur, pointsDetectedPrev;
    // 初始化这些数据
    for(int k=0; k<predictor.num_parts(); ++k)
    {
        pointsPrev.push_back(cv::Point2f(0,0));
        points.push_back(cv::Point2f(0,0));
        pointsDetectedCur.push_back(cv::Point2f(0,0));
        pointsDetectedPrev.push_back(cv::Point2f(0,0));
    }

    // 第一帧没啥子好处理的
    bool isFirstFrame=true;

    // 是否启用光流法
    bool showStabilized = false;


    while(true)
    {
        if(count==0)
          t=cv::getTickCount();

        // 获取图像
        cap >> im;
        // 转为灰度图像
        cv::cvtColor(im, imGray, cv::COLOR_BGR2GRAY);


        cv::resize(im, imSmall, cv::Size(), 1.0/RESIZE_SCALE, 1.0/RESIZE_SCALE);

        // 将图片从opencv格式更改为dlib格式
        dlib::cv_image<bgr_pixel> cimSmall(imSmall);
        dlib::cv_image<bgr_pixel> cimg(im);

        if(count % SKIP_FRAMES ==0)
        {
            faces=detector(cimSmall);
        }

        if(faces.size()<1) //没有检测到脸的时候不处理
            continue;

        // 获取每张脸的特征点
        std::vector<dlib::full_object_detection> shapes;
        for(int i=0;i<faces.size();i++)
        {
            dlib::rectangle r((long)(faces[i].left()*RESIZE_SCALE),(long)(faces[i].top()*RESIZE_SCALE),(long)(faces[i].right()*RESIZE_SCALE),(long)(faces[i].bottom()*RESIZE_SCALE));

            // 找出每张脸的shape
            dlib::full_object_detection shape=predictor(cimg,r);
            shapes.push_back(shape);

            // 循环处理每一个点
            for(int k=0;k<shape.num_parts();k++)
            {
                if(isFirstFrame)
                {
                    pointsPrev[k].x = pointsDetectedPrev[k].x = shape.part(k).x();
                    pointsPrev[k].y = pointsDetectedPrev[k].y = shape.part(k).y();
                }
                else
                {
                    pointsPrev[k]=points[k];
                    pointsDetectedPrev[k]=pointsDetectedCur[k];
                }

                points[k].x = pointsDetectedCur[k].x=shape.part(k).x();
                points[k].y = pointsDetectedCur[k].y=shape.part(k).y();
            }

            // 计算Sigma的值
            if(eyeDistanceNotCalculated)
            {
                eyeDistance=interEyeDistance(shape);
                winSize = cv::Size(2 * int(eyeDistance/4) + 1,  2 * int(eyeDistance/4) + 1);
                eyeDistanceNotCalculated = false;
                dotRadius = eyeDistance > 100 ? 3 : 2;
                sigma = eyeDistance * eyeDistance / 400;
            }

            // 计算图像金字塔
            cv::buildOpticalFlowPyramid(imGray, imGrayPyr, winSize, maxLevel);
            // 使用光流法预测关键点的位置
            cv::calcOpticalFlowPyrLK(imGrayPrevPyr, imGrayPyr, pointsPrev, points, status, err, winSize, maxLevel, termcrit, 0, 0.0001);

            // 获取权重
            for(int k=0;k<shape.num_parts();k++)
            {
                double n= norm(pointsDetectedPrev[k]-pointsDetectedCur[k]);
                double alpha = exp(-n*n/(sigma*1.0));
                std::cout<<"sigma="<<sigma<<std::endl;
                std::cout<<"alpha="<<alpha<<std::endl;
                points[k]=(1-alpha)*pointsDetectedCur[k]+alpha*points[k];
            }

            if(showStabilized)
            {
                renderFacePoints(im, points, cv::Scalar(255,0,255),dotRadius);
            }
            else
            {
                renderFacePoints(im, points, cv::Scalar(0,255,255), dotRadius);
            }

        }


        // 切换
        char key=cv::waitKey(1);
        if(key==27)
        {
            return EXIT_SUCCESS;
        }
        else if(key==32) // 点击空格键切换
        {
            showStabilized = !showStabilized;
        }

        // 准备下一帧
        imPrev=im.clone();
        imGrayPrev=imGray.clone();
        imGrayPrevPyr=imGrayPyr;
        imGrayPyr= std::vector<cv::Mat>();

        isFirstFrame=false;

        count ++;

        if(count ==100)
        {
            t=((double)cv::getTickCount() -t)/cv::getTickFrequency();
            fps=100.0/t;
            count=0;
        }

        // 显示帧率
        cv::putText(im, cv::format("fps %.2f", fps), cv::Point(50, size.height-50),cv::FONT_HERSHEY_COMPLEX, 1.5, cv::Scalar(0,0,255),3);
        cv::imshow(winName, im);
    }

    cap.release();
    cv::destroyAllWindows();

    return EXIT_SUCCESS;
}

File: classes03/python/webcamStabilizedFLD.py
#
# @brief 提升Facial Landmark Detector的识别速度
# 
# @file fastWebcamFLD.py
# @author 张晓民
# @date 2018-08-24
#
import math
import sys
import cv2
import dlib
import numpy as np
from renderFace import renderFace

PREDICTOR_PATH = "../../common/resources/shape_predictor_68_face_landmarks.dat"
RESIZE_HEIGHT = 480
NUM_FRAMES_FOR_FPS = 100
SKIP_FRAMES = 2

# 计算眼睛外眼角的距离
def interEyeDistance(predict):
    leftEyeLeftCorner = (predict[36].x, predict[36].y)
    rightEyeRightCorner = (predict[45].x, predict[45].y)
    distance = cv2.norm(np.array(rightEyeRightCorner)-np.array(leftEyeLeftCorner))
    distance = int(distance)
    return distance



winName="Stablizing Facial Landmark Detector"
cv2.namedWindow(winName, cv2.WINDOW_NORMAL)

# 初始化摄像头对象
cap = cv2.VideoCapture(0)

# 检查是否成功打开摄像头
if(cap.isOpened() is False):
    print("打开摄像头失败")
    sys.exit()

print("摄像头打开成功")

winSize = 101
maxLevel = 10
fps=30.0

# 获取第一针图像
ret, imPrev=cap.read()

# 将当前帧转化为灰度图像
imGrayPrev = cv2.cvtColor(imPrev, cv2.COLOR_BGR2GRAY)

# 加载脸部探测器和形状评估器
detector = dlib.get_frontal_face_detector()
predictor = dlib.shape_predictor(PREDICTOR_PATH)

# 初始化参数
points = [] #当前的landmark坐标
pointsPrev = [] #之前的landmark的坐标
pointsDetectedCur = [] #当前检测到的landmark的坐标
pointsDetectedPrev = [] #上一阵检测到的landmark的坐标

eyeDistanceNotCalculated = True
eyeDistance = 0
isFirstFrame = True

fps = 10
showStabilized = False
count = 0

while(True):
    if(count == 0):
        t = cv2.getTickCount()
    
    # 获取一帧数据
    ret, im = cap.read()
    # 将获取的数据转换为灰度图像
    imGray = cv2.cvtColor(im,cv2.COLOR_BGR2GRAY)
    height = im.shape[0]
    IMAGE_RESIZE = float(height)/RESIZE_HEIGHT
    # 缩小图片使得追踪速度加快
    imSmall = cv2.resize(im,None, fx=1.0/IMAGE_RESIZE,fy=1.0/IMAGE_RESIZE,interpolation=cv2.INTER_LINEAR)

    # 跳帧达到加速的效果
    if(count%SKIP_FRAMES == 0):
        faces = detector(imSmall,0)
    
    # 检测是否检测到了脸部
    if len(faces) == 0:
        print("未检测到脸部")
    
    # 扫描每一张脸，检测器关键点
    else:
        for i in range(0,len(faces)):
            print("第%d张脸：" % (i+1))

            newRect = dlib.rectangle(int(faces[i].left()*IMAGE_RESIZE),
                int(faces[i].top()*IMAGE_RESIZE),
                int(faces[i].right()*IMAGE_RESIZE),
                int(faces[i].bottom()*IMAGE_RESIZE))
            
            # 第一帧要单独处理
            if(isFirstFrame):
                pointsPrev=[]
                pointsDetectedPrev=[]
                [pointsPrev.append((p.x, p.y)) for p in predictor(im, newRect).parts()]
                [pointsDetectedPrev.append((p.x, p.y)) for p in predictor(im, newRect).parts()]
            else:
                pointsPrev = []
                pointsDetectedPrev = []
                pointsPrev = points
                pointsDetectedPrev = pointsDetectedCur
            
            # pointsDetectedCur来储存用facial landmark detector检测的结果
            points = []
            pointsDetectedCur = []
            [points.append((p.x, p.y)) for p in predictor(im, newRect).parts()]
            [pointsDetectedCur.append((p.x, p.y)) for p in predictor(im, newRect).parts()]

            # 将numpy转化为float array
            pointsArr = np.array(points, np.float32)
            pointsPreArr = np.array(pointsPrev, np.float32)

            # 如果眼睛的距离
            if eyeDistanceNotCalculated:
                eyeDistance = interEyeDistance(predictor(im, newRect).parts())
                print("眼睛的距离:%d" % eyeDistance)
                eyeDistanceNotCalculated = False
            
            if eyeDistance > 100:
                dotRadius = 3
            else:
                dotRadius = 2
            
            sigma = eyeDistance*eyeDistance/400
            s = 2*int(eyeDistance/4)+1

            # 设置光流法参数
            lk_params = dict(winSize=(s,s), maxLevel=5, criteria = (cv2.TERM_CRITERIA_COUNT | cv2.TERM_CRITERIA_EPS, 20, 0.03))

            pointsArr, status, err = cv2.calcOpticalFlowPyrLK(imGrayPrev, imGray, pointsPreArr, pointsArr, **lk_params)

            sigma = 100

            # 转换为float
            pointsArrFloat = np.array(points, np.float32)

            # 转化为list
            points = pointsArrFloat.tolist()

            for k in range(0, len(predictor(im, newRect).parts())):
                d=cv2.norm(np.array(pointsDetectedPrev[k]) - np.array(pointsDetectedCur[k]))
                alpha = math.exp(-d*d/sigma)
                points[k]=(1-alpha)*np.array(pointsDetectedCur[k])+alpha*np.array(points[k])
            
            # 绘制关键特征点
            if showStabilized is True:
                for p in points:
                    cv2.circle(im, (int(p[0]), int(p[1])), dotRadius, (255,0,0), -1)
            else:
                for p in pointsDetectedCur:
                    cv2.circle(im, (int(p[0]), int(p[1])), dotRadius, (0,0,255), -1)
            
            isFirstFrame = False
            count = count +1

            # 计算FPS
            if (count == NUM_FRAMES_FOR_FPS):
                t=(cv2.getTickCount()-t)/cv2.getTickFrequency()
                fps = NUM_FRAMES_FOR_FPS/t
                count = 0
                isFirstFrame=True

            cv2.putText(im, "{0:.2f}-fps".format(fps), (50, im.shape[0]-50), cv2.FONT_HERSHEY_COMPLEX, 1.5, (0,0,255), 3, cv2.LINE_AA)
            cv2.imshow(winName, im)
            key = cv2.waitKey(25) & 0xFF

            if key == 32:
                showStabilized = not showStabilized
            
            if key == 27:
                sys.exit()
            

cv2.destroyAllWindows()
cap.release()



